x/incentives/keeper/proposals.go

```diff
@@ -0,0 +1,118 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/tharsis/evmos/x/incentives/types"
+)
+
+// RegisterIncentive creates an incentive for a contract
+func (k Keeper) RegisterIncentive(
+	ctx sdk.Context,
+	contract common.Address,
+	allocations sdk.DecCoins,
+	epochs uint32,
+) (*types.Incentive, error) {
+
+	// check if the Incentives are globally enabled
+	params := k.GetParams(ctx)
+	if !params.EnableIncentives {
+		return nil, sdkerrors.Wrap(
+			types.ErrInternalIncentive,
+			"incentives are currently disabled by governance",
+		)
+	}
+
+	// check if the incentive is already registered
+	if k.IsIncentiveRegistered(ctx, contract) {
+		return nil, sdkerrors.Wrapf(
+			types.ErrInternalIncentive,
+			"incentive already registered: %s", contract,
+		)
+	}
+
+	// check if the balance is > 0 for coins other than the mint denomination
+	mintDenom := k.mintKeeper.GetParams(ctx).MintDenom
+	for _, al := range allocations {
+		if al.Denom != mintDenom {
+			if !k.bankKeeper.HasSupply(ctx, al.Denom) {
+				return nil, sdkerrors.Wrapf(
+					sdkerrors.ErrInvalidCoins,
+					"base denomination '%s' cannot have a supply of 0", al.Denom,
+				)
+			}
+		}
+
+		// check if each allocation is below the allocation limit
+		if al.Amount.GT(params.AllocationLimit) {
+			return nil, sdkerrors.Wrapf(
+				types.ErrInternalIncentive,
+				"allocation for denom '%s' (%v) cannot be above allocation limmit '%v' - ", al.Denom, al.Amount, params.AllocationLimit,
+			)
+		}
+	}
+
+	// check if the sum of all allocations for each denom (current + proposed) is
+	// < 100%
+	incentives := k.GetAllIncentives(ctx)
+	if len(incentives) != 0 {
+		currentAllocations := make(map[string]sdk.Dec)
+
+		k.IterateIncentives(
```

Comment:
I think we could track the total allocated incentives (%) in a new variable on the store. This would save us the cost of the iterating every contract during registration

Reply:
this would be also useful for clients (eg: "oh look there's only 50% allocated, let's create a new proposal to incentivize our contract too!")

Reply:
Added AllocationMeter to solve this denom (key) -> sdk.Int (value)